src/bin/customer/pay.rs

```diff
@@ -22,91 +23,137 @@ use super::{connect, database, Command};
+                &self.label,
+                session_key,
+                chan,
+                payment_amount,
+            ),
         )
         .await
+        .context("Payment timed out while updating channel status")?
         .context("Failed to complete pay protocol")?;
 
-        // Receive the response note (i.e. the fulfillment of the service)
-        let (response_note, chan) = chan
-            .recv()
+        tokio::time::timeout(config.approval_timeout(), receive_service(chan))
             .await
-            .context("Failed to receive response note")?;
+            .context("Payment timed out when receiving service")??;
 
-        // Close the communication channel: we are done communicating with the merchant
-        chan.close();
+        Ok(())
+    }
+}
 
-        // Print the response note on standard out
-        if let Some(response_note) = response_note {
-            eprintln!(
-                "Payment succeeded with response from merchant: \"{}\"",
-                response_note
-            );
-        } else {
-            eprintln!("Payment succeeded with no concluding response from merchant");
-        }
+/// Parse the payment amount specified on the command line.
+fn parse_payment_amount(amount: Amount) -> Result<PaymentAmount, anyhow::Error> {
+    // Convert the payment amount appropriately
+    let minor_units: i64 = amount.try_into_minor_units().ok_or_else(|| {
+        anyhow::anyhow!("Payment amount invalid for currency or out of range for channel")
+    })?;
 
-        Ok(())
+    (if minor_units < 0 {
+        PaymentAmount::pay_customer
+    } else {
+        PaymentAmount::pay_merchant
+    })(minor_units.abs() as u64)
+    .context("Payment amount out of range")
+}
+
+/// Set up the communication channel with the merchant.
+async fn open_session(
+    database: &dyn QueryCustomer,
+    config: &Config,
+    channel_name: &ChannelName,
+) -> Result<(SessionKey, Chan<pay::Pay>), anyhow::Error> {
+    // Look up the address and current local customer state for this merchant in the database
+    let address = database
+        .channel_address(channel_name)
+        .await
+        .context("Failed to look up channel address in local database")?;
+
+    // Connect and select the Pay session
+    let (session_key, chan) = connect(config, &address).await?;
+    let chan = chan
+        .choose::<2>()
+        .await
+        .context("Failed selecting pay session with merchant")?;
+
+    Ok((session_key, chan))
+}
+
+/// Request approval for the payment request from the merchant, aborting the session if it is not
+/// granted.
+async fn request_payment(
+    config: &Config,
+    chan: Chan<pay::Pay>,
+    payment_amount: PaymentAmount,
+    payment_note: Option<Note>,
+) -> Result<Chan<pay::CustomerStartPayment>, anyhow::Error> {
+    // Read the contents of the note, if any
+    let note = payment_note
+        .unwrap_or_else(|| zeekoe::customer::cli::Note::String(String::from("")))
```

Comment:
Could we `impl Default for Note` and `unwrap_or_default()`?

Reply:
Ooh, nice.